hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -28,45 +48,102 @@ object InferPType {
         while(vit.hasNext) {
           val value = vit.next()
           InferPType(value, env)
-          assert(head.pType2 == value.pType2)
+          assert(head.pType2 isOfType value.pType2)
         }
 
         head.pType2
       }
       case Ref(name, _) => env.lookup(name)
       case In(_, t) => PType.canonical(t)
-      case MakeArray(_, t) => PType.canonical(t)
-      case MakeStream(_, t) => PType.canonical(t)
+        // TODO: not sure if preference is to continue using canonical here
+        // I could imagine a collection of non-canonical ptypes
+      case MakeArray(irs, _) => {
+        val it = irs.iterator
+        val head = it.next()
+        InferPType(head, env)
+        var allRequired = head.pType2.required
+
+        while(it.hasNext) {
+          val irElem = it.next()
+
+          InferPType(irElem, env)
+          assert(head.pType2 isOfType irElem.pType2)
+
+          if(allRequired == true && irElem.pType2.required == false) {
+            allRequired = false
+          }
+        }
+
+        PArray(head.pType2, allRequired)
```

Comment:
This should be `PArray(head.pType2.setRequired(allRequired), true)`.

Reply:
I think PArray(head.pType2.setRequired(allRequired), allRequired)?

Reply:
The array itself *is* required regardless of the requiredness of children! `[None, None, None] is not None` is `True`.

Reply:
but this isn't really correct -- the full type of the children needs to be traversed. If you have:
```
MakeArray(
  Ref(x, +PStruct{x: +PInt32}),
  Ref(y, PStruct{x: PInt32}))
```

Then this algorithm will return `PStruct{x: +PInt32}`, which is wrong

Reply:
I think I'm confused then about what requiredeness means. I thought we were going to use this to set the missing bits on the array, i.e that setting required = true means that none of the children are missing, and require = false means that some (or all) are.

Reply:
The `If` needs to include recursive requiredness unification as well!

The `cnsq` and `altr` (or MakeArray children) can have different memory layouts, and these need to be unified.

Reply:
> I think I'm confused then about what requiredeness means. I thought we were going to use this to set the missing bits on the array, i.e that setting required = true means that none of the children are missing, and require = false means that some (or all) are.

This isn't quite right. We're using this to determine *whether we need to allocate an array of missing bits at all*. If we know at compile time that the elements can't be missing, there is no need to check the missing bits at run time, therefore there is no need for missing bits to exist at run time.

Note that it's perfectly possible for an `PArray(PInt32(false))` to have no missing elements. Having `required=true` just means we were able to prove at compile time that *every possible* array we might construct will have no missing elements (and therefore an array of missing bits would be redundant).

Reply:
> We're using this to determine whether we need to allocate an array of missing bits at all. 

Your language is more precise, this is what I meant; whether we need this header.


> Note that it's perfectly possible for an PArray(PInt32(false)) to have no missing elements. Having required=true just means we were able to prove at compile time that every possible array we might construct will have no missing elements (and therefore an array of missing bits would be redundant).

I understand this with regard to PArray(PType, true). I'm not sure how we can prove at compile time that PArray(PType, true) if we can't prove PType.required == true, meaning when is `PArray.elementType.required != PArray.required`? I'm confused how the array's requiredeness isn't a direct consequence of the requiredeness of all of it's children, unless the `required` attribute on the `PArray` is an indicator that we require the array to be allocated (even if full of only-missing data), which was mentioned in another comment, and which seems to be a polymorphic use of this attribute.

Reply:
Yeah, the array's requiredness isn't about whether or not it needs to allocate missing bits for its elements, but whether or not the array has to exist/be non-missing. If I have `arrayType.required == true`, that only says that the array itself can never be missing. If I say `arrayType.elementType.required == true`, then I don't have to track whether the elements of the array are missing---and hence the array can avoid storing a header for missing bits. Does that make sense?

Reply:
Here's a simple example: `MakeArray(NA(TInt32()))`. The strongest type we can give this is `PArray(PInt32(false), true)`. Clearly the elements can be missing: we know statically that there is only one element, and that it is missing! But just as clearly we know statically that the array itself is not missing: we know statically that it is exactly the array `[NA]`, which is not the same as the array `NA`.

More generally, if `x` is a value we got from somewhere with type `PInt32(false)`, i.e. we can't prove that it is definitely not missing, then `MakeArray(x)` should have type `PArray(PInt32(false), true)`, because the array itself is definitely not missing (we just constructed it!).

Reply:
Yep!